File: agenkit-java/src/main/java/io/agenkit/core/Message.java
package io.agenkit.core;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Immutable message exchanged between agents and users.
 */
public final class Message {

    private static final Set<String> ALLOWED_ROLES = Set.of(
            "user", "assistant", "system", "tool", "agent");
    private static final int MAX_CONTENT_BYTES = 16 * 1024 * 1024; // 16 MB
    private static final int MAX_METADATA_KEYS = 100;
    private static final int MAX_KEY_LENGTH = 50;
    private static final int MAX_ROLE_LENGTH = 20;

    private final String role;
    private final Object content;
    private final Map<String, Object> metadata;
    private final Instant timestamp;

    public Message(String role, Object content, Map<String, Object> metadata, Instant timestamp) {
        this.role = role;
        this.content = content;
        this.metadata = metadata != null
                ? Collections.unmodifiableMap(new HashMap<>(metadata))
                : Collections.emptyMap();
        this.timestamp = timestamp != null ? timestamp : Instant.now();
    }

    /** Factory: create a simple user/assistant message. */
    public static Message of(String role, String content) {
        return new Message(role, content, Collections.emptyMap(), Instant.now());
    }

    /** Returns content as a String, or empty string if null. */
    public String contentString() {
        if (content == null) return "";
        if (content instanceof String s) return s;
        return content.toString();
    }

    /** Returns a copy of this message with an additional metadata entry. */
    public Message withMetadata(String key, Object value) {
        Map<String, Object> newMeta = new HashMap<>(this.metadata);
        newMeta.put(key, value);
        return new Message(role, content, newMeta, timestamp);
    }

    /** Validates this message and returns it; throws IllegalArgumentException on failure. */
    public Message validate() {
        if (role == null || role.isEmpty()) {
            throw new IllegalArgumentException("message role cannot be empty");
        }
        if (role.length() > MAX_ROLE_LENGTH) {
            throw new IllegalArgumentException(
                    "message role exceeds maximum length of " + MAX_ROLE_LENGTH
                            + " characters (got " + role.length() + ")");
        }
        if (!ALLOWED_ROLES.contains(role)) {
            throw new IllegalArgumentException(
                    "invalid message role: " + role
                            + ". Must be one of: user, assistant, system, tool, agent");
        }

        String contentStr = contentString();
        int contentSize = contentStr.getBytes(StandardCharsets.UTF_8).length;
        if (contentSize > MAX_CONTENT_BYTES) {
            throw new IllegalArgumentException(
                    "message content exceeds maximum size of " + MAX_CONTENT_BYTES
                            + " bytes (got " + contentSize + " bytes)");
        }

        if (metadata.size() > MAX_METADATA_KEYS) {
            throw new IllegalArgumentException(
                    "message metadata exceeds maximum of " + MAX_METADATA_KEYS
                            + " keys (got " + metadata.size() + ")");
        }
        for (Map.Entry<String, Object> entry : metadata.entrySet()) {
            String key = entry.getKey();
            if (key.length() > MAX_KEY_LENGTH) {
                throw new IllegalArgumentException(
                        "metadata key '" + key.substring(0, Math.min(20, key.length()))
                                + "...' exceeds maximum length of " + MAX_KEY_LENGTH
                                + " characters (got " + key.length() + ")");
            }
        }
        return this;
    }

    public String getRole() { return role; }
    public Object getContent() { return content; }
    public Map<String, Object> getMetadata() { return metadata; }
    public Instant getTimestamp() { return timestamp; }

    @Override
    public String toString() {
        return "Message{role='" + role + "', content='" + contentString() + "'}";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Message m)) return false;
        return role.equals(m.role)
                && java.util.Objects.equals(content, m.content)
                && metadata.equals(m.metadata)
                && timestamp.equals(m.timestamp);
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(role, content, metadata, timestamp);
    }
}

File: agenkit-cs/src/Agenkit/Core/Message.cs
namespace Agenkit.Core;

/// <summary>
/// Represents a message exchanged between agents or with users.
/// </summary>
public record Message(
    string Role,
    object? Content,
    IReadOnlyDictionary<string, object>? Metadata = null,
    DateTimeOffset? Timestamp = null)
{
    private static readonly HashSet<string> AllowedRoles = new(StringComparer.Ordinal)
    {
        "user", "assistant", "system", "tool", "agent"
    };

    private const int MaxContentBytes = 16 * 1024 * 1024; // 16MB
    private const int MaxMetadataKeys = 100;
    private const int MaxKeyLength = 50;
    private const int MaxRoleLength = 20;

    /// <summary>
    /// Returns the message content as a string.
    /// </summary>
    public string ContentString() => Content switch
    {
        string s => s,
        null => "",
        _ => Content.ToString() ?? ""
    };

    /// <summary>
    /// Creates a new message with the given role and content.
    /// </summary>
    public static Message NewMessage(string role, string content) =>
        new(role, content, new Dictionary<string, object>(), DateTimeOffset.UtcNow);

    /// <summary>
    /// Validates the message and returns it. Throws ArgumentException if invalid.
    /// </summary>
    public Message Validate()
    {
        if (string.IsNullOrEmpty(Role))
            throw new ArgumentException("message role cannot be empty");

        if (Role.Length > MaxRoleLength)
            throw new ArgumentException(
                $"message role exceeds maximum length of {MaxRoleLength} characters (got {Role.Length})");

        if (!AllowedRoles.Contains(Role))
            throw new ArgumentException(
                $"invalid message role: {Role}. Must be one of: user, assistant, system, tool, agent");

        // Content size validation
        var contentSize = Content switch
        {
            string s => System.Text.Encoding.UTF8.GetByteCount(s),
            null => 0,
            _ => System.Text.Encoding.UTF8.GetByteCount(Content.ToString() ?? "")
        };

        if (contentSize > MaxContentBytes)
            throw new ArgumentException(
                $"message content exceeds maximum size of {MaxContentBytes} bytes (got {contentSize} bytes)");

        if (Metadata is not null)
        {
            if (Metadata.Count > MaxMetadataKeys)
                throw new ArgumentException(
                    $"message metadata exceeds maximum of {MaxMetadataKeys} keys (got {Metadata.Count})");

            foreach (var (key, value) in Metadata)
            {
                if (key.Length > MaxKeyLength)
                    throw new ArgumentException(
                        $"metadata key '{key[..Math.Min(20, key.Length)]}...' exceeds maximum length of {MaxKeyLength} characters (got {key.Length})");

                var valueStr = value?.ToString() ?? "";
                var valueSize = System.Text.Encoding.UTF8.GetByteCount(valueStr);
                if (valueSize > MaxContentBytes)
                    throw new ArgumentException(
                        $"metadata value for key '{key}' exceeds maximum size of {MaxContentBytes} bytes (got {valueSize} bytes)");
            }
        }

        return this;
    }

    /// <summary>
    /// Returns a new message with additional metadata.
    /// </summary>
    public Message WithMetadata(string key, object value)
    {
        var meta = Metadata is not null
            ? new Dictionary<string, object>(Metadata) { [key] = value }
            : new Dictionary<string, object> { [key] = value };
        return this with { Metadata = meta };
    }
}
